alternator/ttl.cc

```diff
@@ -113,4 +142,494 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+// to split the expiration scanner's work between all the shards of the
+// system.
+// When some of the nodes in the system are down, nobody will be expiring
+// tokens owned by them, so we will need to separately detect when this
+// happens and let living nodes take over the work that a dead node cannot
+// do.
+// FIXME: need to decide how to choose primary ranges in multi-DC setup!
+// We could call get_primary_ranges_within_dc() below instead of get_primary_ranges().
+class token_ranges_owned_by_this_shard {
+    schema_ptr _s;
+    // _token_ranges will contain a list of token ranges owned by this node.
+    // We'll further need to split each such range to the pieces owned by
+    // the current shard, using _intersecter.
+    dht::token_range_vector _token_ranges;
+    dht::token_range_vector::const_iterator _range_it;
+    std::optional<dht::selective_token_range_sharder> _intersecter;
+public:
+    // FIXME: add a constructor to start iteration in the middle, to continue
+    // a scan after reboot. A simpler option is to begin the scan at a random
+    // position.
+    token_ranges_owned_by_this_shard(database& db, schema_ptr s)
+        :  _s(s)
+        , _token_ranges(db.find_keyspace(s->ks_name()).
+            get_effective_replication_map()->get_primary_ranges(
+            utils::fb_utilities::get_broadcast_address()))
+        , _range_it(_token_ranges.cbegin())
+    {}
+
+    // Return the next token_range owned by this shard, or nullopt when the
+    // iteration ends.
+    std::optional<dht::token_range> next() {
+        // We may need three or more iterations in the following loop if a
+        // vnode doesn't intersect with the given shard at all (such a small
+        // vnode is unlikely, but possible). The loop cannot be infinite
+        // because each iteration of the loop advances _range_it.
+        for (;;) {
+            if (_intersecter) {
+                std::optional<dht::token_range> ret = _intersecter->next();
+                if (ret) {
+                    return ret;
+                }
+                // done with this range, go to next one
+                ++_range_it;
+                _intersecter = std::nullopt;
+            }
+            if (_range_it == _token_ranges.cend()) {
+                return std::nullopt;
+            }
+            _intersecter.emplace(_s->get_sharder(), *_range_it, this_shard_id());
+        }
+    }
+
+    // Same as next(), just return a partition_range instead of token_range
+    std::optional<dht::partition_range> next_partition_range() {
+        std::optional<dht::token_range> ret = next();
+        if (ret) {
+            return dht::to_partition_range(*ret);
+        } else {
+            return std::nullopt;
+        }
+    }
+};
+
+// Scan data in a list of token ranges in one table, looking for expired
+// items and deleting them.
+// Because of issue #9167, partition_ranges must have a single partition
```

Comment:
btw - fix for #9167 landed on master; still, it's fine to address it in a follow-up to this series to avoid an endless loop of review

Reply:
Yes, I saw that, but 1. you're right about the endless loop, and 2. I'm not sure I trust this fix :-( I didn't see a unit test or any explanation of how it was tested, and now I no that nothing in the code uses multiple disconnected ranges, so I worry about being the first user. I'll think about this again later.